src/main/kube-auth-proxy.ts

```diff
@@ -58,52 +62,58 @@ export class KubeAuthProxy {
-    const args = [
-      "proxy",
-      "-p", "0",
-      "--kubeconfig", `${this.cluster.kubeConfigPath}`,
-      "--context", `${this.cluster.contextName}`,
-      "--accept-hosts", this.acceptHosts,
-      "--reject-paths", "^[^/]"
-    ];
-
-    if (process.env.DEBUG_PROXY === "true") {
-      args.push("-v", "9");
+  public async run(): Promise<void> {
+    if (!this.proxyProcess) {
+      const proxyBin = await this.kubectl.getPath();
+      const args = [
+        "proxy",
+        "-p", "0",
+        "--kubeconfig", `${this.cluster.kubeConfigPath}`,
+        "--context", `${this.cluster.contextName}`,
+        "--accept-hosts", this.acceptHosts,
+        "--reject-paths", "^[^/]"
+      ];
+
+      if (process.env.DEBUG_PROXY === "true") {
+        args.push("-v", "9");
+      }
+      logger.debug(`spawning kubectl proxy with args: ${args}`);
+
+      this.proxyProcess = spawn(proxyBin, args, { env: this.env, });
+      this.proxyProcess.on("error", (error) => {
+        this.sendIpcLogMessage({ data: error.message, error: true });
+        this.exit();
+      });
+
+      this.proxyProcess.on("exit", (code) => {
+        this.sendIpcLogMessage({ data: `proxy exited with code: ${code}`, error: code > 0 });
+        this.exit();
+      });
+
+      this.proxyProcess.stderr.on("data", (data) => {
+        this.lastError = this.parseError(data.toString());
+        this.sendIpcLogMessage({ data: data.toString(), error: true });
+      });
+
+      this._port = await getPortFrom(this.proxyProcess.stdout, {
+        lineRegex: startingServeRegex,
+        onFind: () => this.sendIpcLogMessage({ data: "Authentication proxy started\n" }),
+      });
+
+      this.proxyProcess.stdout.on("data", (data: any) => {
+        this.sendIpcLogMessage({ data: data.toString() });
+      });
+
+      await waitUntilUsed(this.port, 500, 10000);
+    
+      this.ready = true;
     }
-    logger.debug(`spawning kubectl proxy with args: ${args}`);
-
-    this.proxyProcess = spawn(proxyBin, args, { env: this.env, });
-    this.proxyProcess.on("error", (error) => {
-      this.sendIpcLogMessage({ data: error.message, error: true });
-      this.exit();
-    });
-
-    this.proxyProcess.on("exit", (code) => {
-      this.sendIpcLogMessage({ data: `proxy exited with code: ${code}`, error: code > 0 });
-      this.exit();
-    });
-
-    this.proxyProcess.stderr.on("data", (data) => {
-      this.lastError = this.parseError(data.toString());
-      this.sendIpcLogMessage({ data: data.toString(), error: true });
-    });
-
-    this._port = await getPortFrom(this.proxyProcess.stdout, {
-      lineRegex: startingServeRegex,
-      onFind: () => this.sendIpcLogMessage({ data: "Authentication proxy started\n" }),
-    });
-
-    this.proxyProcess.stdout.on("data", (data: any) => {
-      this.sendIpcLogMessage({ data: data.toString() });
-    });
-
-    return waitUntilUsed(this.port, 500, 10000);
+    
+    return this.whenReady;
```

Comment:
Since you removed the `await run()` shouldn't this not return this promise?

Reply:
Not sure I understand. I removed the `await` from the `run()` call in `ensureServer()` just because `ensureServer()` has to finally await the `whenReady` anyway. I could leave it as `await run()`. Does it matter?

`run()` has to return this promise, for when it gets called when `this.proxyProcess` is defined already, right?

Reply:
Yes I see why you removed the `await`. I think that is correct. However, because you did that this getter is creating a new promise which is then wrapped into the upper promise and then thrown away because no one await it.

I think it would be better to just not return anything. Even better it would be nice if the promise inside `run()` didn't leak at all out of it. So that the return type was `void`. This would make it inline with a `start()` function.

The corollary of that could be an `ensureStart()` which calls start and then awaits the `whenReady` getter promise. But that might be a bit too much.